crates/ui/src/components/scrollbar.rs

```diff
@@ -59,49 +53,63 @@ impl ScrollableHandle {
     }
     fn set_offset(&self, point: Point<Pixels>) {
         let base_handle = match self {
-            ScrollableHandle::Uniform(handle) => &handle.0.borrow().base_handle,
-            ScrollableHandle::NonUniform(handle) => &handle,
+            ListScrollableHandle::Uniform(handle) => &handle.0.borrow().base_handle,
+            ListScrollableHandle::NonUniform(handle) => &handle,
         };
         base_handle.set_offset(point);
     }
     fn offset(&self) -> Point<Pixels> {
         let base_handle = match self {
-            ScrollableHandle::Uniform(handle) => &handle.0.borrow().base_handle,
-            ScrollableHandle::NonUniform(handle) => &handle,
+            ListScrollableHandle::Uniform(handle) => &handle.0.borrow().base_handle,
+            ListScrollableHandle::NonUniform(handle) => &handle,
         };
         base_handle.offset()
     }
     fn viewport(&self) -> Bounds<Pixels> {
         let base_handle = match self {
-            ScrollableHandle::Uniform(handle) => &handle.0.borrow().base_handle,
-            ScrollableHandle::NonUniform(handle) => &handle,
+            ListScrollableHandle::Uniform(handle) => &handle.0.borrow().base_handle,
+            ListScrollableHandle::NonUniform(handle) => &handle,
         };
         base_handle.bounds()
     }
 }
-impl From<UniformListScrollHandle> for ScrollableHandle {
+
+impl From<UniformListScrollHandle> for ListScrollableHandle {
     fn from(value: UniformListScrollHandle) -> Self {
         Self::Uniform(value)
     }
 }
 
-impl From<ScrollHandle> for ScrollableHandle {
+impl From<ScrollHandle> for ListScrollableHandle {
     fn from(value: ScrollHandle) -> Self {
         Self::NonUniform(value)
     }
 }
 
+#[derive(Debug)]
+pub struct ContentSize {
+    pub size: Size<Pixels>,
+    pub scroll_adjustment: Option<Point<Pixels>>,
+}
+
+pub trait ScrollableHandle: Clone + 'static {
+    fn content_size(&self) -> Option<ContentSize>;
+    fn set_offset(&self, point: Point<Pixels>);
+    fn offset(&self) -> Point<Pixels>;
+    fn viewport(&self) -> Bounds<Pixels>;
+}
+
 /// A scrollbar state that should be persisted across frames.
 #[derive(Clone, Debug)]
-pub struct ScrollbarState {
+pub struct ScrollbarState<T: ScrollableHandle> {
```

Comment:
What'd you say to using a trait object for `scroll_handle`? That way you wouldn't have to carry a generic argument around. Plus, you could drop `Clone` requirement on your ScrollableHandle, as you could store the trait object in an `Arc`.

Reply:
Yes, initially, I did try that with `Box` and `boxed_clone`. It even worked, but I went with the generic route for no strong reason. I'll update it as you mentioned.